### PR TITLE
Allow creation of internet shortcut files

### PR DESCRIPTION
**Pull request by h-h-h-h**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

This pull request solves an issue I posted earlier (#16402). You can read about my reasoning there. (It got at least one supporter very quickly, despite of being closed already!)

However, the way I solved the issue was not by adding a new `--format`, but to write internet shortcut files in the same way as you can, e.g., create a `.description` file. I added these switches (from `README.md`):

```
    --write-link                     Write an internet shortcut file, depending
                                     on the current platform (.url/.webloc/
                                     .desktop). The URL may be cached by the OS.
    --write-url-link                 Write a Windows internet shortcut file
                                     (.url). Note that the OS caches the URL
                                     based on the file path.
    --write-webloc-link              Write a macOS internet shortcut file
                                     (.webloc)
    --write-desktop-link             Write a Linux internet shortcut file
                                     (.desktop)
```

Example usage for a YouTube channel with 30 videos (found via [incognitube.com](https://www.incognitube.com/)):

```
youtube-dl --write-link --skip-download --download-archive .youtube-dl-download-archive --ignore-errors https://www.youtube.com/user/LAMBORGHINIDAB23369/videos
```

This creates internet shortcuts to URLs of the form `https://www.youtube.com/watch?v=IXrSW__RrEo`.

When trying to download an RSS feed like https://www.blubrry.com/feeds/smallandmighty.xml (21 items; retrieved from https://itunes.apple.com/gb/podcast/the-small-and-mighty-podcast/id1228157605 via [getrssfeed.com](http://getrssfeed.com/)), the URLs inside the internet shortcut files are of the form: `http://media.blubrry.com/smallandmighty/content.blubrry.com/smallandmighty/Kristina_Karlsson_Podcast_final.mp3` -- that is, direct media files.

The behavior of `--download-archive` was changed to a small degree. From `README.md`:

```
    --download-archive FILE          Download only videos not listed in the
                                     archive file. Record the IDs of all
                                     downloaded videos in it. When the switches
                                     --write-link (or similar) and
                                     --skip-download are used additionally, the
                                     IDs will also be recorded, even though
                                     nothing was actually downloaded.
```

I ran the tests I changed successfully under:

- Windows 10, Python 3.5.4
- Ubuntu, Python 2.7.12

`--write-link` also created the correct filetype under each OS.

---

The function `to_high_limit_path()` from `utils.py` that this PR introduces allows youtube-dl to get rid of Windows' `MAX_PATH` limitations. It should be used everywhere before `io.open()`. **(Create an issue?)** At this point, it is just used when creating internet shortcut files.

Video titles with emojis are supported by `to_high_limit_path()`. Among other videos, this was tested with:

```
youtube-dl --write-link --skip-download https://www.youtube.com/watch?v=o-fuq5GcLAI
```

---

`flake8` still reports issues. See dedicated comment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -176,7 +176,11 @@ Alternatively, refer to the [developer instructions](#developer-instructions) fo
                                      age
```
Comment (dstftw) [inline README.md]: This is auto generated. Do not modify it.

File: youtube_dl/YoutubeDL.py [hunk 2]
```diff
@@ -1407,6 +1419,8 @@ def process_video_result(self, info_dict, download=True):
             raise ExtractorError('Missing "id" field in extractor result')
         if 'title' not in info_dict:
             raise ExtractorError('Missing "title" field in extractor result')
+        if 'webpage_url' not in info_dict:
+            raise ExtractorError('Missing "webpage_url" field in extractor result. Should have been augmented with it.')
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: No.

File: youtube_dl/YoutubeDL.py [hunk 3]
```diff
@@ -1836,7 +1850,56 @@ def ensure_dir_exists(path):
...
+            if not _write_link_file('url',     URL_LINK_TEMPLATE,     '\r\n', embed_filename=False): return
+        if webloc_link:
+            if not _write_link_file('webloc',  WEBLOC_LINK_TEMPLATE,  '\n',   embed_filename=False): return
+        if desktop_link:
+            if not _write_link_file('desktop', DESKTOP_LINK_TEMPLATE, '\n',   embed_filename=True ): return
+
+        if self.params.get('skip_download', False):
+            # Regarding the download archive, consider internet shortcut creation in conjunction with the `--skip-download` switch as everything the user wants. (See also help for the`--download-archive` switch.)
+            if url_link or webloc_link or desktop_link:
+                self.record_download_archive(info_dict)
```
Comment (dstftw) [inline youtube_dl/YoutubeDL.py]: Download archive behavior must not change, it must only take place after success of the actual download and post processing.

**Comment by h-h-h-h** [comment]
@dstftw, would you review again, please?

**Pull request closed without merging**